drivers/display/ssd1322.c

```diff
@@ -0,0 +1,353 @@
+}
+
+static inline int ssd1322_write_data(const struct device *dev, uint8_t *buf, size_t len)
+{
+	const struct ssd1322_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+
+	mipi_desc.buf_size = len;
+	mipi_desc.width = len * PIXELS_IN_BYTE;
+	mipi_desc.height = 1;
+	mipi_desc.pitch = mipi_desc.width;
+	return mipi_dbi_write_display(config->mipi_dev, &config->dbi_config, buf, &mipi_desc,
+				      PIXEL_FORMAT_MONO01);
+}
+
+static int ssd1322_blanking_on(const struct device *dev)
+{
+	return ssd1322_write_command(dev, SSD1322_BLANKING_ON, NULL, 0);
+}
+
+static int ssd1322_blanking_off(const struct device *dev)
+{
+	return ssd1322_write_command(dev, SSD1322_BLANKING_OFF, NULL, 0);
+}
+
+static int ssd1322_write(const struct device *dev, const uint16_t x, const uint16_t y,
+			 const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct ssd1322_config *config = dev->config;
+	size_t buf_len;
+	int ret;
+	uint8_t cmd_data[2];
+	int32_t pixel_count = desc->width * desc->height;
+	const uint8_t *buf_iter;
+	static uint8_t conversion_buf[CONFIG_SSD1322_BUFFER_SIZE];
+
+	if (desc->pitch < desc->width) {
+		LOG_ERR("Pitch is smaller than width");
+		return -EINVAL;
+	}
+
+	buf_len = MIN(desc->buf_size, desc->height * desc->width / 8);
+	if (buf == NULL || buf_len == 0U) {
+		LOG_ERR("Display buffer is not available");
+		return -EINVAL;
+	}
+
+	if (desc->pitch > desc->width) {
+		LOG_ERR("Unsupported mode");
+		return -EINVAL;
+	}
+
+	LOG_DBG("x %u, y %u, pitch %u, width %u, height %u, buf_len %u", x, y, desc->pitch,
+		desc->width, desc->height, buf_len);
+
+	cmd_data[0] = config->column_offset + (x >> 2) * config->segments_per_pixel;
+	cmd_data[1] =
+		config->column_offset + ((x + desc->width) >> 2) * config->segments_per_pixel - 1;
+	ret = ssd1322_write_command(dev, SSD1322_SET_COLUMN_ADDR, cmd_data, 2);
+	if (ret < 0) {
+		return ret;
+	}
+
+	cmd_data[0] = y;
+	cmd_data[1] = y + desc->height - 1;
+	ret = ssd1322_write_command(dev, SSD1322_SET_ROW_ADDR, cmd_data, 2);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = ssd1322_write_command(dev, SSD1322_ENABLE_RAM_WRITE, NULL, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/*
+	 * Controller uses 4-bit grayscale format, so one pixel is represented by 4 bits.
+	 * Zephyr's display API does not support this format, so I am using 1-bit b/w mode and
+	 * convert each 1-bit pixel to 1111 or 0000.
+	 */
+	__ASSERT(CONFIG_SSD1322_BUFFER_SIZE / config->segments_per_pixel / 2 > 0,
+		 "Pixel conversion buffer too small");
+
+	buf_iter = buf;
+	while (pixel_count > 0) {
+		uint16_t in_idx = 0;
+		uint16_t seg_idx = 0;
+		uint8_t color;
+
+		while (in_idx < pixel_count && seg_idx / 2 < sizeof(conversion_buf)) {
+			in_idx = seg_idx / config->segments_per_pixel;
+			color = (buf_iter[in_idx / 8] & BIT(in_idx % 8)) ? 0xF : 0;
+			conversion_buf[seg_idx / 2] =
+				(seg_idx % 2 == 0) ? color
+						   : ((color << 4) | conversion_buf[seg_idx / 2]);
+			seg_idx++;
+		}
```

Comment:
It does not look nice, maybe you can move it to a separate e.g. mono01_to_grayscale conversion function.

Reply:
Agreed. I restructured the whole conversion logic to be slightly more readable.